sdks/python/apache_beam/testing/synthetic_pipeline.py

```diff
@@ -116,6 +118,149 @@ def process(self, element):
+      result = list(
+          OffsetRange(restriction[0], restriction[1]).split(
+              offsets_per_split, offsets_per_split // 2))
+      return [(x.start, x.stop) for x in result]
+
+  def restriction_size(self, element, restriction):
+    if self._size_estimate_override is not None:
+      return self._size_estimate_override
+    element_size = len(element) if isinstance(element, str) else 1
+    return (restriction[1] - restriction[0]) * element_size
+
+
+def getSyntheticSDFStep(per_element_delay_sec=0,
+                        per_bundle_delay_sec=0,
+                        output_records_per_input_record=1,
+                        output_filter_ratio=0,
+                        initial_splitting_num_bundles=8,
+                        initial_splitting_uneven_chunks=False,
+                        disable_liquid_sharding=False,
+                        size_estimate_override=None,):
+  """ A function which returns a SyntheticSDFStep with given parameters. """
+
+  class SyntheticSDFStep(beam.DoFn):
+    """A SplittableDoFn of which behavior can be controlled through prespecified
+       parameters.
+    """
+
+    def __init__(self, per_element_delay_sec_arg, per_bundle_delay_sec_arg,
+                 output_filter_ratio_arg):
+      if per_element_delay_sec_arg and per_element_delay_sec_arg < 1e-3:
+        raise ValueError(
+            'Per element sleep time must be at least 1e-3. '
+            'Received: %r', per_element_delay_sec_arg)
+      self._per_element_delay_sec = per_element_delay_sec_arg
+      self._per_bundle_delay_sec = per_bundle_delay_sec_arg
+      self._output_filter_ratio = output_filter_ratio_arg
+
+    def start_bundle(self):
+      self._start_time = time.time()
+
+    def finish_bundle(self):
+      # The target is for the enclosing stage to take as close to as possible
+      # the given number of seconds, so we only sleep enough to make up for
+      # overheads not incurred elsewhere.
+      to_sleep = self._per_bundle_delay_sec - (
+          time.time() - self._start_time)
+
+      # Ignoring sub-millisecond sleep times.
+      if to_sleep >= 1e-3:
+        time.sleep(to_sleep)
+
+    def process(self,
+                element,
+                restriction_tracker=beam.DoFn.RestrictionParam(
+                    SyntheticSDFStepRestrictionProvider(
+                        output_records_per_input_record,
+                        initial_splitting_num_bundles,
+                        initial_splitting_uneven_chunks,
+                        disable_liquid_sharding,
+                        size_estimate_override))):
+      filter_element = False
+      if self._output_filter_ratio > 0:
+        if np.random.random() < self._output_filter_ratio:
+          filter_element = True
+
+      for k in range(*restriction_tracker.current_restriction()):
+        if not restriction_tracker.try_claim(k):
+          return
+
+        if self._per_element_delay_sec > 0:
```

Comment:
In the synthetic pipeline map step, the per element delay is how much we should wait in time per **input** element and not per output that we produce.

I think we should divide the wait time by num elements we are going to emit and perform these smaller sleeps to make progress through the SDF, wdyt?